Tests/AsyncQueueTests/SemaphoreTests.swift

```diff
@@ -0,0 +1,144 @@
+        /// A counter that will only be accessed from within the `systemUnderTest`'s context
+        let unsafeCounter = UnsafeCounter()
+
+        for _ in 1...iterationCount {
+            await systemUnderTest.enqueueAndCount(using: unsafeCounter) { systemUnderTest in
+                let didSuspend = await systemUnderTest.wait()
+                XCTAssertTrue(didSuspend)
+
+                return { systemUnderTest in
+                    // Signal that the suspended wait call above has resumed.
+                    // This signal allows us to `wait()` for all of these enqueued `wait()` tasks to have completed later in this test.
+                    systemUnderTest.signal()
+                }
+            }
+        }
+
+        // Loop one fewer than iterationCount.
+        for _ in 1..<iterationCount {
+            await systemUnderTest.execute { systemUnderTest in
+                systemUnderTest.signal()
+            }
+        }
+
+        await systemUnderTest.execute { systemUnderTest in
+            // Give each suspended `wait` task an opportunity to resume (if they were to resume, which they won't) before we check the count.
+            for _ in 1...iterationCount {
+                await Task.yield()
+            }
+
+            // The count will still be zero each time because we have executed one more `wait` than `signal` calls.
+            let completedCountedTasks = unsafeCounter.countedTasksCompleted
+            XCTAssertEqual(completedCountedTasks, 0)
+
+            // Signal one last time, enabling all of the original `wait` calls to resume.
+            systemUnderTest.signal()
+
+            for _ in 1...iterationCount {
+                // Wait for all enqueued `wait`s to have completed and signaled their completion.
+                await systemUnderTest.wait()
+            }
+
+            let tasksCompleted = unsafeCounter.countedTasksCompleted
+            XCTAssertEqual(iterationCount, tasksCompleted)
+        }
+    }
+
+    func test_wait_doesNotSuspendIfSignalCalledFirst() async {
+        await systemUnderTest.signal()
+        let didSuspend = await systemUnderTest.wait()
+        XCTAssertFalse(didSuspend)
+    }
+
+    // MARK: Private
+
+    private var systemUnderTest = Semaphore()
+}
+
+// MARK: - Semaphore Extension
+
+private extension Semaphore {
+    /// Enqueues an asynchronous task. This method suspends the caller until the asynchronous task has begun, ensuring ordered execution of enqueued tasks.
+    /// - Parameter task: A unit of work that returns work to execute after the task completes and the count is incremented.
+    func enqueueAndCount(using counter: UnsafeCounter, _ task: @escaping @Sendable (isolated Semaphore) async -> ((isolated Semaphore) -> Void)?) async {
+        // Await the start of the soon-to-be-enqueued `Task` with a continuation.
+        await withCheckedContinuation { continuation in
+            // Re-enter the actor context but don't wait for the result.
```

Comment:
To make sure I follow... the actor context is the semaphore's context? And the reason that this task is on that context is because this method is isolated to the semaphore by the nature of it being an extension on an actor?

Reply:
Curious to get your thoughts on this ^ when you have a moment @dfed . Tagging you here in case you missed this one.

Reply:
I did miss this!

> the actor context is the semaphore's context

yes!

> the reason that this task is on that context is because this method is isolated to the semaphore by the nature of it being an extension on an actor

yes!

Thoughts on changing "the actor context" to "the semaphore context"?

Reply:
> Thoughts on changing "the actor context" to "the semaphore context"?

Yeah I think that would increase clarity 👍

Reply:
4f38bfe